codalab/worker/file_util.py

```diff
@@ -171,11 +184,192 @@ def do_unzip(filename):
+
+    def __init__(self, buffer, cleanup_fn, cleanup_args):
+        self._buffer = buffer
+        self._cleanup_fn = cleanup_fn
+        self._cleanup_args = cleanup_args
+
+    def close(self, *args, **kwargs):
+        return self._buffer.close(*args, **kwargs)
+
+    @property
+    def closed(self):
+        return self._buffer.closed
+
+    def fileno(self, *args, **kwargs):
+        return self._buffer.fileno(*args, **kwargs)
+
+    def flush(self, *args, **kwargs):
+        return self._buffer.flush(*args, **kwargs)
+
+    def isatty(self, *args, **kwargs):
+        return self._buffer.isatty(*args, **kwargs)
+
+    def readable(self, *args, **kwargs):
+        return self._buffer.readable(*args, **kwargs)
+
+    def readline(self, *args, **kwargs):
+        return self._buffer.readline(*args, **kwargs)
+
+    def readlines(self, *args, **kwargs):
+        return self._buffer.readlines(*args, **kwargs)
+
+    def seek(self, *args, **kwargs):
+        return self._buffer.seek(*args, **kwargs)
+
+    def seekable(self, *args, **kwargs):
+        return self._buffer.seekable(*args, **kwargs)
+
+    def tell(self, *args, **kwargs):
+        return self._buffer.tell(*args, **kwargs)
+
+    def truncate(self, *args, **kwargs):
+        return self._buffer.truncate(*args, **kwargs)
+
+    def writable(self, *args, **kwargs):
+        return self._buffer.writable(*args, **kwargs)
+
+    def writelines(self, *args, **kwargs):
+        return self._buffer.writelines(*args, **kwargs)
+
+    def read(self, *args, **kwargs):
+        return self._buffer.read(*args, **kwargs)
+
+    def readall(self, *args, **kwargs):
+        return self._buffer.readall(*args, **kwargs)
+
+    def readinto(self, *args, **kwargs):
+        return self._buffer.readinto(*args, **kwargs)
+
+    def write(self, *args, **kwargs):
+        return self._buffer.write(*args, **kwargs)
+
+    def __del__(self, *args, **kwargs):
+        self._cleanup_fn(*self._cleanup_args)
+        return self._buffer.__del__(*args, **kwargs)
+
+
+def open_file(file_path, mode='r'):
+    """
+    Opens the file indicated by the given file path. Can be in a directory.
+    """
+    linked_bundle_path = parse_linked_bundle_url(file_path)
+    if (
+        linked_bundle_path.uses_beam
+        and linked_bundle_path.is_archive
+        and linked_bundle_path.archive_subpath
+    ):
+        # If file path is a .tar.gz file on Azure, open the specified path within the archive.
+        tf, _ = open_indexed_tar_gz_file(linked_bundle_path.bundle_path)
+
+        isdir = lambda finfo: finfo.type == tarfile.DIRTYPE
+
+        fpath = "/" + linked_bundle_path.archive_subpath
+        finfo = tf.getFileInfo(fpath)
+        if finfo is None:
+            raise FileNotFoundError(fpath)
+        if isdir(finfo):
+            from codalab.worker.download_util import compute_target_info_beam_descendants_flat
+
+            # If streaming a folder within an Azure bundle, we need to download its contents,
```

Comment:
This is potentially a big bottleneck if this is happening on the server...should be able to stream it somehow, right?

Reply:
Okay, I was able to stream this (see https://github.com/codalab/codalab-worksheets/blob/beam-du/codalab/worker/tar_subdir_stream.py and https://github.com/codalab/codalab-worksheets/blob/beam-du/codalab/worker/tar_file_stream.py) -- it's a bit complicated but is working well.